### PR TITLE
CORDA-1939 Regression: Liquibase fails to migrate existing pre-liquibase database

### PR DESCRIPTION
**Pull request by szymonsztuka**
Recently, Liquibase was added to support schema modification to internal Corda tables.
The issue occurred when for initial Liquibase setup for the existing H2 database, the coda-finance module was not up to date (were missing Liquibase migration scripts) and the migration failed leaving database in inconsistent state.
The fix:
* corda.jar can be manually deployed without coping the matching corda-finance-VERSION.jar.
If  the older finance cordapp version is detected (which doesn't have Liquibase migration scripts) fail node at startup and print message:
"Could not create the DataSource: Detected incompatible corda-finance cordapp without database migration scripts, replace the existing corda-finance-VERSION.jar with the latest one."
Since coda-finance is an optional cordapp,the  presence of Liquibase scripts is only checked if corda-finance-VERSION.jar is present in cordapps folder.

Unrelated change to the bug fix:
* Allow to start using Liquibase from any point of 4.0-SNAPSHOT before Liquibase was introduced (not only from 3.0/3.X versions) - 2 database changes introduced after 3.2 but before Liquibase are now conditional database changes

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/schema/NodeSchemaService.kt [hunk 1]
```diff
@@ -69,7 +69,7 @@ class NodeSchemaService(private val extraSchemas: Set<MappedSchema> = emptySet()
      if (includeNotarySchemas) mapOf(Pair(NodeNotaryV1, SchemaOptions())) else emptyMap()
 
     fun internalSchemas() = requiredSchemas.keys + extraSchemas.filter { schema -> // when mapped schemas from the finance module are present, they are considered as internal ones
-        schema::class.simpleName == "net.corda.finance.schemas.CashSchemaV1" || schema::class.simpleName == "net.corda.finance.schemas.CommercialPaperSchemaV1" }
+        schema::class.qualifiedName == "net.corda.finance.schemas.CashSchemaV1" || schema::class.qualifiedName == "net.corda.finance.schemas.CommercialPaperSchemaV1" }
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/services/schema/NodeSchemaService.kt]: How did the code ever work before if this conditional was incorrect?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/SchemaMigration.kt [hunk 2]
```diff
@@ -119,29 +125,32 @@ class SchemaMigration(
...
+                                "migration/cash.changelog-v1.xml"))
+
+                    if (schemas.any { schema -> schema.migrationResource == "commercial-paper.changelog-master" })
+                        preV4Baseline.addAll(listOf("migration/commercial-paper.changelog-init.xml",
+                                "migration/commercial-paper.changelog-v1.xml"))
+
+                    if (schemas.any { schema -> schema.migrationResource == "node-notary.changelog-master" })
+                        preV4Baseline.addAll(listOf("migration/node-notary.changelog-init.xml",
+                                "migration/node-notary.changelog-v1.xml",
+                                "migration/vault-schema.changelog-pkey.xml"))
```
Comment (josecoll) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/SchemaMigration.kt]: Seems rather odd that the Vault Schema is updated differently for a Notary node.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/SchemaMigration.kt [hunk 3]
```diff
@@ -106,10 +106,16 @@ class SchemaMigration(
 
     /** For existing database created before verions 4.0 add Liquibase support - creates DATABASECHANGELOG and DATABASECHANGELOGLOCK tables and mark changesets are executed. */
     private fun migrateOlderDatabaseToUseLiquibase(existingCheckpoints: Boolean): Boolean {
+        //workaround to detect that if Corda finance module is in use then the most recent version with Liquibase migration scripts was deployed
+        if (schemas.any { schema ->
+                    (schema::class.qualifiedName == "net.corda.finance.schemas.CashSchemaV1" || schema::class.qualifiedName == "net.corda.finance.schemas.CommercialPaperSchemaV1")
```
Comment (josecoll) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/SchemaMigration.kt]: I'm not a fan of using FQCN in strings.
Perhaps use `CashSchemaV1::class.qualifiedName`

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/SchemaMigration.kt (hunk 2 above)
Comment (szymonsztuka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/SchemaMigration.kt]: Indeed this is wrong, vault-schema is already included before, so removing it.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/SchemaMigration.kt (hunk 3 above)
Comment (szymonsztuka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/SchemaMigration.kt]: The problem here (and in the other place) is that node-api doesn't know about finance module, but SchemaMigration needs to know that MappedSchemas from finance are tracked by Liquibase.

**Comment by szymonsztuka** [comment]
yes, it introduces  tighter coupling  between node and finance but not compile one, it comes down to marking MappedSchemas from optional Finance Cordapp to migrate via Liquibase, there could be other ways to do it  though (e.g. a subclass `InternalMappedSchema` )

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/SchemaMigration.kt (hunk 3 above)
Comment (josecoll) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/SchemaMigration.kt]: But why is `SchemaMigration` in the `node-api` module if it is solely used internally by the node?

Comment (szymonsztuka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/SchemaMigration.kt]: This is to be consistent with Enterprise repo.

**Pull request merged**